pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -140,51 +138,89 @@ func getWorkloadEntryHandler(c *ServiceEntryStore) func(model.Config, model.Conf
 		// only when services have changed - otherwise, just push endpoint updates.
-		fp := true
-		if event == model.EventUpdate {
+		var addedSvcs, deletedSvcs, updatedSvcs, unchangedSvcs []*model.Service
+
+		switch event {
+		case model.EventUpdate:
 			// This is not needed, update should always have old populated, but just in case.
 			if old.Spec != nil {
 				os := convertServices(old)
-				fp = servicesChanged(os, cs) || selectorChanged(old, curr)
+				if selectorChanged(old, curr) {
+					// Consider all services are updated.
+					mark := make(map[host.Name]*model.Service, len(cs))
+					for _, svc := range cs {
+						mark[svc.Hostname] = svc
+						updatedSvcs = append(updatedSvcs, svc)
+					}
+					for _, svc := range os {
+						if _, f := mark[svc.Hostname]; !f {
+							updatedSvcs = append(updatedSvcs, svc)
+						}
+					}
+				} else {
+					addedSvcs, deletedSvcs, updatedSvcs, unchangedSvcs = servicesDiff(cs, os)
+				}
 			} else {
+				unchangedSvcs = cs
 				log.Warnf("Spec is not available in the old service entry during update, proceeding with full push %v", old)
 			}
-		}
-
-		// If service entry is deleted, cleanup endpoint shards for services.
-		if event == model.EventDelete {
+		case model.EventDelete:
+			deletedSvcs = cs
+			// If service entry is deleted, cleanup endpoint shards for services.
 			for _, svc := range cs {
 				c.XdsUpdater.SvcUpdate(c.Cluster(), string(svc.Hostname), svc.Attributes.Namespace, event)
 			}
+		case model.EventAdd:
+			addedSvcs = cs
+		default:
+			unchangedSvcs = cs
 		}
 
+		for _, svcs := range [][]*model.Service{addedSvcs, deletedSvcs, updatedSvcs} {
+			for _, svc := range svcs {
+				configsUpdated[model.ConfigKey{
+					Kind:      model.ServiceEntryKind,
+					Name:      string(svc.Hostname),
+					Namespace: svc.Attributes.Namespace}] = struct{}{}
+			}
+		}
+
+		willFullPush := len(configsUpdated) > 0
+		refreshIndexes := willFullPush // If will trigger full-push, lazy-update instance index.
+
 		// Recomputing the index here is too expensive - lazy build when it is needed.
 		c.changeMutex.Lock()
-		c.refreshIndexes = fp // Only recompute indexes if services have changed.
+		c.refreshIndexes = refreshIndexes // Only recompute indexes if services have changed.
 		c.changeMutex.Unlock()
 
-		if fp {
-			pushReq := &model.PushRequest{
-				Full:              true,
-				NamespacesUpdated: map[string]struct{}{curr.Namespace: {}},
-				ConfigsUpdated:    map[resource.GroupVersionKind]map[string]struct{}{serviceEntryKind: {}},
-				Reason:            []model.TriggerReason{model.ServiceUpdate},
+		if len(unchangedSvcs) > 0 && !willFullPush {
+			// If will do full-push, leave the edsUpdate to that.
+			// XXX We should do edsUpdate for all unchangedSvcs since we begin to calculate service
+			// data according to this "configsUpdated" and thus remove the "!willFullPush" condition.
+			instances := convertInstances(curr, unchangedSvcs)
+			if !refreshIndexes {
```

Comment:
nit: this is always true

Reply:
The `refreshIndexes` and `willFullPush` has overlapped value. I used two vars instead of a single one to express separated business semantics though we can derive one from another. 

How do you think of this kind of work?